### ISSUE TITLE
Move scrapy.utils.deprecate to separate package

### ISSUE THREAD

**Issue by Digenis** [scrapy/scrapy#2194]
Among the pypi packages with a purpose similar to
`scrapy.utils.deprecate.create_deprecated_class`,
scrapy's implementation stands out
being the only one that warns when subclassing.

I think it deserves its own package
where further development can continue
without inflating scrapy's codebase.

@dangra, you have the name in pypi reserved.
Do you have a plan to create a package?

**Comment by kmike** [comment]
Me and @dangra wanted to create this package quite a long time ago; the idea was indeed to move scrapy deprecation utilities to a separate package, likely with a different API. We never found time to do that, but it is still in plans :)

**Comment by kmike** [comment]
I think we should reduce the scope of it a bit, and create a package just with `create_deprecated_class`. What do you think @Gallaecio? You had to copy this code to https://github.com/scrapinghub/web-poet recently, right?

**Comment by Gallaecio** [comment]
I agree that it deserves a place in a package that assists with code deprecation, instead of keeping a copy of it in every package where it is used. I am OK with reducing its scope, although I recall finding it lacking for some scenarios.

However, rather than creating a package for it, I wonder if we can find some active package of deprecation utilities instead that would accept a pull request to add something like this.

**Comment by Gallaecio** [comment]
Looking at GitHub stars, https://github.com/tantale/deprecated seems like the most popular package on this front, followed by https://github.com/briancurtin/deprecation and https://github.com/Borda/pyDeprecate.

**Comment by wRAR** [comment]
3.13 has it in the stdlib as `@warnings.deprecated`. I wonder if there are/will be backports of it.

**Comment by wRAR** [comment]
> A runtime implementation of the @deprecated decorator is available in the typing-extensions library since version 4.5.0.

**Comment by kmike** [comment]
warnings.deprecated doesn't seem to handle inheritance the same way our implementation does.

**Comment by dangra** [comment]
it could be a good upstream contribution tho

**Comment by wRAR** [comment]
https://peps.python.org/pep-0702/ (which added it) may be useful in the context of this discussion, I haven't read it though. It mentions https://github.com/tantale/deprecated, also mentioned by Adrian above, not sure if that one can be used instead of the Scrapy code or lacks something.

**Comment by AdrianAtZyte** [comment]
https://github.com/scrapy/formerly
